grove/arbitrary_state/arbitrary_state.py

```diff
@@ -0,0 +1,214 @@
+"""Class for generating a program that can generate an arbitrary quantum state
+
+- http://140.78.161.123/digital/\
+2016_ismvl_logic_synthesis_quantum_state_generation.pdf
+- https://arxiv.org/pdf/quant-ph/0407010.pdf
+
+"""
+import numpy as np
+import pyquil.quil as pq
+from pyquil.api import SyncConnection
+from pyquil.gates import *
+
+
+def get_uniformly_controlled_rotation_matrix(k):
+    """
+    Returns the matrix represented by :math:`M_{ij}` in arXiv:quant-ph/0407010.
+
+    This matrix converts the angles of :math:`k`-fold uniformly
+    controlled rotations to the angles of the efficient gate decomposition.
+
+    :param int k: number of control qubits
+    :return: the matrix :math:`M_{ij}`
+    :rtype: 2darray
+    """
+    M = np.full((2 ** k, 2 ** k), 2 ** -k)
+    for i in xrange(2 ** k):
+        g_i = i ^ (i >> 1)  # Gray code for i
+        for j in xrange(2 ** k):
+            M[i, j] *= (-1) ** (bin(j & g_i).count("1"))
+    return M
+
+
+def get_cnot_control_positions(k):
+    """
+    Returns a list of positions for the controls of the CNOTs used when
+    decomposing uniformly controlled rotations, as outlined in
+    arXiv:quant-ph/0407010.
+
+    It is assumed that the target is the zeroth qubit
```

Comment:
`k` is actually an `int`, which is just the number of control qubits. I've updated the docstring; it really doesn't matter what the target qubit is labelled, I just wanted to make it clear that this function assumes the controls are labelled 1, 2, ..., k in order from bottom to top and the indices in the return list are labelled with this convention.